src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/graphs/AdaptiveChainPruner.java

```diff
@@ -29,41 +33,105 @@ public AdaptiveChainPruner(final double initialErrorProbability, final double lo
+        verticesToProcess.add(maxWeightChain.getFirstVertex());
+        verticesToProcess.add(maxWeightChain.getLastVertex());
+
+        // look for vertices with two incoming or two outgoing chains (plus one outgoing or incoming for a total of 3 or more) with good log odds to seed the subgraph of good vertices
+        // the logic here is that a high-multiplicity error chain A that branches into a second error chain B and a continuation-of-the-original-error chain A'
+        // may have a high log odds for A'.  However, only in the case of true variation will  multiple branches leaving the same vertex have good log odds.
+        vertexToSeedableChains.keySet().stream()
+                .filter(v -> vertexToSeedableChains.get(v).size() > 2)
+                .forEach(verticesToProcess::add);
+
+        final Set<V> processedVertices = new HashSet<>();
+        final Set<Path<V,E>> goodChains = new HashSet<>();
+
+        // starting from the high-confidence seed vertices, grow the "good" subgraph along chains with above-threshold log odds,
+        // discovering good chains as we go.
+        while (!verticesToProcess.isEmpty()) {
+            final V vertex = verticesToProcess.poll();
+            processedVertices.add(vertex);
+            for (final Path<V,E> outgoingChain : vertexToGoodOutgoingChains.get(vertex)) {
+                goodChains.add(outgoingChain);
+                if(!processedVertices.contains(outgoingChain.getLastVertex())) {
+                    verticesToProcess.add(outgoingChain.getLastVertex());
+                }
+            }
 
-        chainLogOdds.forEach((chain, lod) -> {
-            if (lod < logOddsThreshold) {
-                result.add(chain);
+            for (final Path<V,E> incomingChain : vertexToGoodIncomingChains.get(vertex)) {
+                goodChains.add(incomingChain);
+                if(!processedVertices.contains(incomingChain.getFirstVertex())) {
+                    verticesToProcess.add(incomingChain.getFirstVertex());
+                }
             }
-        });
+        }
+
+        final Set<Path<V,E>> errorChains = chains.stream().filter(c -> !goodChains.contains(c)).collect(Collectors.toSet());
 
-        chains.stream().filter(c -> isChainPossibleVariant(c, graph))
-                .sorted(Comparator.comparingDouble((ToDoubleFunction<Path<V, E>>) chainLogOdds::get)
-                        .reversed().thenComparingInt(Path::length))
+        // add non-error chains to error chains if maximum number of variants has been exceeded
+        chains.stream()
+                .filter(c -> !errorChains.contains(c))
+                .filter(c -> isChainPossibleVariant(c, graph))
+                .sorted(Comparator.comparingDouble(c -> Math.min(chainLogOdds.get(c).getLeft(), chainLogOdds.get(c).getLeft())).reversed())
                 .skip(maxUnprunedVariants)
-                .forEach(result::add);
+                .forEach(errorChains::add);
```

Comment:
You know, it occurs to me that the `maxUnprunedVariants` limit here probably doesn't make sense. If you think about it there is not a 1:1 mapping of chains to variants after pruning. Its very possible for a "variant" to be a single chain in the simple case or possible correspond to a few distinct "chains" because there could be forked paths with low LOD that get pruned that end up splitting that one good "chain" into a few smaller ones. Furthermore it seems possible that a perfectly good variant has part a part of its chain removed at this stage as a result resulting in 1 or potentially 2 dangling ends after pruning and deleting the connection. I don't suspect this happens often enough to worry about here though. Worth keeping in mind for future changes though.

Reply:
You have a good point, and it's worth fixing now. The number of variants associated with a vertex is the number of good outgoing chains minus one.  This works even where there is one good incoming chain, one good outgoing chain, and one bad outgoing chain.  Thus we have

number of variants in graph = sum_vertices (max(number of outgoing chains - 1, 0))

After computing this number of variants we could prune good chains, starting from those with the worst log odds, until the number of variants is low enough.  Every time we prune a good chain whose first vertex has at least one other good outgoing chain we have eliminated one variant from the graph.

Reply:
Done.